File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/AddCategoryFragment.java
package com.example.gemainventory.ui.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;
import com.example.gemainventory.R;
import com.example.gemainventory.api.RetrofitClient;
import com.example.gemainventory.model.CategoriaDto;
import com.google.android.material.textfield.TextInputEditText;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AddCategoryFragment extends Fragment {

    private TextInputEditText etName, etDesc;
    private Button btnSave;
    private int categoryId = 0; // 0 = New
    private CategoriaDto categoriaEditable;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_add_category, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        etName = view.findViewById(R.id.et_category_name);
        etDesc = view.findViewById(R.id.et_category_desc);
        btnSave = view.findViewById(R.id.btn_save_category);

        view.findViewById(R.id.btn_back).setOnClickListener(v -> NavHostFragment.findNavController(this).popBackStack());
        
        // CHECK EDIT MODE
        if (getArguments() != null && getArguments().containsKey("category_id")) {
             categoryId = getArguments().getInt("category_id");
             String name = getArguments().getString("category_name");
             String desc = getArguments().getString("category_desc");
             
             etName.setText(name);
             if(desc != null) etDesc.setText(desc);
             
             ((android.widget.TextView)view.findViewById(R.id.tv_header_title)).setText("Editar Categoría");
             btnSave.setText("Actualizar Categoría");
        }

        btnSave.setOnClickListener(v -> guardarCategoria());
    }

    private void guardarCategoria() {
        String nombre = etName.getText().toString().trim();
        String descripcion = etDesc.getText().toString().trim();
        
        if (nombre.isEmpty()) {
            etName.setError("Requerido");
            return;
        }

        SharedPreferences prefs = requireActivity().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE);
        String userId = prefs.getString("user_id", null);

        btnSave.setEnabled(false);
        
        if (categoryId == 0) {
            // CREATE
            CategoriaDto nueva = new CategoriaDto(0, nombre, descripcion);
            Call<CategoriaDto> call = RetrofitClient.INSTANCE.getInstance().crearCategoria(nueva, userId);
            call.enqueue(new Callback<CategoriaDto>() {
                @Override
                public void onResponse(Call<CategoriaDto> call, Response<CategoriaDto> response) {
                    if (response.isSuccessful()) {
                        Toast.makeText(getContext(), "Categoría guardada", Toast.LENGTH_SHORT).show();
                        NavHostFragment.findNavController(AddCategoryFragment.this).popBackStack();
                    } else {
                        btnSave.setEnabled(true);
                        Toast.makeText(getContext(), "Error del servidor", Toast.LENGTH_SHORT).show();
                    }
                }
                @Override
                public void onFailure(Call<CategoriaDto> call, Throwable t) {
                    btnSave.setEnabled(true);
                    Toast.makeText(getContext(), "Error de red", Toast.LENGTH_SHORT).show();
                }
            });
        } else {
            // UPDATE
            CategoriaDto edit = new CategoriaDto(categoryId, nombre, descripcion);
            Call<CategoriaDto> call = RetrofitClient.INSTANCE.getInstance().actualizarCategoria(categoryId, edit);
            call.enqueue(new Callback<CategoriaDto>() {
                @Override
                public void onResponse(Call<CategoriaDto> call, Response<CategoriaDto> response) {
                    if (response.isSuccessful()) {
                        Toast.makeText(getContext(), "Categoría actualizada", Toast.LENGTH_SHORT).show();
                        NavHostFragment.findNavController(AddCategoryFragment.this).popBackStack();
                    } else {
                        btnSave.setEnabled(true);
                        Toast.makeText(getContext(), "Error:" + response.code(), Toast.LENGTH_SHORT).show();
                    }
                }
                @Override
                public void onFailure(Call<CategoriaDto> call, Throwable t) {
                    btnSave.setEnabled(true);
                    Toast.makeText(getContext(), "Error de red", Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/AddCategoryFragment.kt
package com.example.gemainventory.ui.settings

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.Label
import androidx.compose.runtime.*
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.example.gemainventory.api.RetrofitClient
import com.example.gemainventory.model.CategoriaDto
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AddCategoryFragment : Fragment() {

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
        val isDarkMode = prefs.getBoolean("DarkMode", false)
        val userId = prefs.getString("user_id", null)

        val args = arguments
        val categoryId = args?.getInt("category_id") ?: 0
        val initialName = args?.getString("category_name") ?: ""
        val initialDesc = args?.getString("category_desc") ?: ""

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                var name by remember { mutableStateOf(initialName) }
                var desc by remember { mutableStateOf(initialDesc) }
                var isLoading by remember { mutableStateOf(false) }

                val fields = listOf(
                    FormFieldData(
                        value = name,
                        onValueChange = { name = it },
                        label = "Nombre de Categoría",
                        icon = Icons.Default.Label
                    ),
                    FormFieldData(
                        value = desc,
                        onValueChange = { desc = it },
                        label = "Descripción",
                        icon = Icons.Default.Description,
                        singleLine = false
                    )
                )

                GenericAddFormScreen(
                    title = if (categoryId == 0) "Nueva Categoría" else "Editar Categoría",
                    darkTheme = isDarkMode,
                    fields = fields,
                    buttonText = if (categoryId == 0) "Guardar Categoría" else "Actualizar Categoría",
                    isLoading = isLoading,
                    onBackClick = { findNavController().popBackStack() },
                    onSaveClick = {
                        if (name.isBlank()) {
                            Toast.makeText(context, "El nombre es requerido", Toast.LENGTH_SHORT).show()
                        } else {
                            isLoading = true
                            saveCategory(categoryId, name, desc, userId) { success ->
                                isLoading = false
                                if (success) findNavController().popBackStack()
                            }
                        }
                    }
                )
            }
        }
    }

    private fun saveCategory(id: Int, name: String, desc: String, userId: String?, onComplete: (Boolean) -> Unit) {
        val dto = CategoriaDto(id, name, desc)
        val call = if (id == 0) {
            RetrofitClient.instance.crearCategoria(dto, userId ?: "")
        } else {
            RetrofitClient.instance.actualizarCategoria(id, dto)
        }

        call.enqueue(object : Callback<CategoriaDto> {
            override fun onResponse(call: Call<CategoriaDto>, response: Response<CategoriaDto>) {
                if (response.isSuccessful) {
                    Toast.makeText(context, "Categoría guardada", Toast.LENGTH_SHORT).show()
                    onComplete(true)
                } else {
                    Toast.makeText(context, "Error: ${response.code()}", Toast.LENGTH_SHORT).show()
                    onComplete(false)
                }
            }

            override fun onFailure(call: Call<CategoriaDto>, t: Throwable) {
                Toast.makeText(context, "Error de red", Toast.LENGTH_SHORT).show()
                onComplete(false)
            }
        })
    }
}
